drgn_tools/list_lru.py

```diff
@@ -0,0 +1,236 @@
+            else:
+                li = lru.node[nid].lru.list
+            for entry in list_for_each_entry(type, li.address_of_(), member):
+                yield entry
+
+
+def list_lru_from_memcg_node_for_each_entry(
+    prog: Program,
+    mindx: IntegerLike,
+    nid: IntegerLike,
+    type: Union[str, Type],
+    lru: Object,
+    member: str,
+) -> Iterator[Object]:
+    """
+    Iterate over the entries in a list_lru by the provided memcg and NUMA node.
+
+    :param prog: Kernel being debugged
+    :param mindx: memcg index.
+    :param nid: NUMA node ID.
+    :param type: Entry type.
+    :param lru: ``struct list_lru *``
+    :param member: Name of list node member in entry type.
+    :return: Iterator of ``type *`` objects.
+    """
+    if node_state(nid, prog["N_ONLINE"]):
+        memcg_aware = 0
+        if has_member(lru, "memcg_aware") and lru.memcg_aware:
+            memcg_aware = 1
+        if has_member(lru, "node"):
+            # no lru.node in uek7 but covered in above test
+            if has_member(lru.node, "memcg_lrus") and lru.node[0].memcg_lrus:
+                memcg_aware = 1
+        if memcg_aware:
+            if has_member(lru, "ext") or has_member(lru, "xa"):
+                # v5.13 (uek7) or newer
+                if has_member(lru, "ext"):
+                    # uek7 has a UEK_KABI_REPLACE of node to ext
+                    xa = lru.ext.xa
+                else:
+                    # uek8
+                    xa = lru.xa
+                # Keep the entries grouped by the NUMA node.
+                memcg = xa_load(xa.address_of_(), mindx)
+                # convert from the void ptr unless it is a NULL
+                if memcg != NULL(prog, "void *"):
+                    memcg = Object(prog, "struct list_lru_memcg *", memcg)
+                    lru_one = memcg.node[nid]
+                    if lru_one.nr_items > 0:
+                        for entry in list_for_each_entry(
+                            type, lru_one.list.address_of_(), member
+                        ):
+                            yield entry
+            else:
+                # Before v5.13
+                # make sure the memcg index is within the legal limits
+                if mindx >= 0 and mindx < prog["memcg_nr_cache_ids"]:
+                    li = lru.node[nid].memcg_lrus.lru[mindx].list
+                    if not list_empty(li.address_of_()):
+                        for entry in list_for_each_entry(
+                            type, li.address_of_(), member
+                        ):
+                            yield entry
+        else:
+            # not lru.memcg_aware
+            if has_member(lru, "ext"):
+                li = lru.ext.node[nid].lru.list
+            else:
+                li = lru.node[nid].lru.list
+            for entry in list_for_each_entry(type, li.address_of_(), member):
+                yield entry
+
+
+def list_lru_kmem_to_memcgidx(prog: Program, kvm: IntegerLike) -> IntegerLike:
```

Comment:
In theory, we could get the numa node id from the page ptr, BUT there are multiple ways that is can be encoded.

Option 1: it is encoded in the vm_page.flags with the offset determined by what else is embedded in the vm_page.flags (section, zone, last cupid)
Option 2: it is  in a special section_to_node_table[] array where the section is encoded in the the vm_page.flags

A check for the section_to_node_table[] is easy enough but the vm_page.flags offsets/width are CONFIG dependent and is less obvious for Drgn  find automatically. From disassembling the code, the hosts that I have, the numa node is:
   (page.flags  >>  0x36)